### PR TITLE
[Spark-20145] Fix range case insensitive bug in SQL 

### PR DESCRIPTION
**Pull request by samelamin**
## What changes were proposed in this pull request?
Range in SQL should be case insensitive

## How was this patch tested?
unit test

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala [hunk 1]
```diff
@@ -105,7 +105,7 @@ object ResolveTableValuedFunctions extends Rule[LogicalPlan] {
 
   override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
     case u: UnresolvedTableValuedFunction if u.functionArgs.forall(_.resolved) =>
-      builtinFunctions.get(u.functionName) match {
+      builtinFunctions.get(u.functionName.toLowerCase) match {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala]: Does the function name is case sensitive in Hive?

**Comment by hvanhovell** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala (hunk 1 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala]: What do you mean? There are no table valued functions in Hive.

**Comment by samelamin** [comment]
I'm not sure to be honest. But even if it is, the inbuilt functions has a
key of "range"

So anything else won't be found which is causing the problem in the first
place


On Thu, 30 Mar 2017 at 23:26, Xiao Li <notifications@github.com> wrote:

*@gatorsmile* commented on this pull request.
------------------------------

In
sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala
<https://github.com/apache/spark/pull/17487#discussion_r109052479>:

> @@ -105,7 +105,7 @@ object ResolveTableValuedFunctions extends Rule[LogicalPlan] {

   override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
     case u: UnresolvedTableValuedFunction if
u.functionArgs.forall(_.resolved) =>
-      builtinFunctions.get(u.functionName) match {
+      builtinFunctions.get(u.functionName.toLowerCase) match {

Does the function name is case sensitive in Hive?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/apache/spark/pull/17487#pullrequestreview-30129926>,
or mute
the thread

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala (hunk 1 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala]: @samelamin you have to make this dependent on the case sensitivity setting of the analyzer. This means you will have to turn this object into case class that takes a SQLConf as its argument.

Comment (samelamin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala]: ok cool ill amend it over the weekend. Do you know how I can run a specific set of tests on SBT? 

im getting garbage collection errors when I try to run using sbt test-only  form http://spark.apache.org/developer-tools.html

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala]: For example, 

```
build/sbt "catalyst/test-only org.apache.spark.sql.catalyst.optimizer.ColumnPruningSuite"
build/sbt -Phive "hive/test-only org.apache.spark.sql.hive.execution.SQLQuerySuite"
```

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala]: @hvanhovell Just did a search. Do we still need to check the case sensitivity setting of the analyzer in this case?

Based on [the Hive doc](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+UDF), 
> All Hive keywords are case-insensitive, including the names of Hive operators and functions.
 
Our regular function names are always case insensitive. See the [code](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala#L74)

Comment (samelamin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala]: @hvanhovell instead of creating a new case class, is there a way I can reuse the UnresolvedTableValuedFunction case class and just add in the SQLConf class?

**Comment by samelamin** [comment]
based on comments from @hvanhovell  I am depending on the case sensitivity setting of the analyser. That said I had to make the functionName a var to change the value to lower case which feels like a code smell to me. I am happy with suggestions to how I can improve on it

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala (hunk 1 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala]: Yeah, @gatorsmile is right. We don't do case sensitive resolution for functions. So you can ignore my comment.

Comment (samelamin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala]: So would you like me to revert the change to the case class?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala [hunk 2]
```diff
@@ -69,9 +70,15 @@ case class UnresolvedInlineTable(
  *   select * from range(10);
  * }}}
  */
-case class UnresolvedTableValuedFunction(functionName: String, functionArgs: Seq[Expression])
+case class UnresolvedTableValuedFunction(conf: SQLConf,
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: @samelamin this is a moot point now, but try to avoid mutable state as much as you can. In this case it would have been better to pass the SQLConf class as a parameter to the `ResolveTableValuedFunction`, and deal with case-sensitivity there.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala (hunk 1 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala]: Yes, sorry about that!

**Comment by samelamin** [comment]
Ok I was planning on doing that but unfortunately all the tests don't go through the resolveTableValuedFuntion but new up an UnTableValued. I totally agree with you with regards to state. What we can do is create a method that calculates the function name inside the unresolvedTableValuedFunction which then has the dependency on the conf. How does that sound?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala (hunk 1 above)
Comment (samelamin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala]: Ok I can revert but how will we test it since the apply method isn't being called by the tests. They new up an unresolvedTableValuedFunction? Unless we hardcore it in unresolvedTableValuedFunction to always be case insensitive then I don't know how else to do it?

**Comment by samelamin** [comment]
Maybe @gatorsmile or @srowen have an idea. How can I test the case insensitive unresolvedTableValuedFunction without using a var for the functionName. I'm open to all ideas folks! introducing a var just for the tests feels inheritely incorrect.

**Comment by samelamin** [comment]
FYI I can squash these commits to clean up the history once we agree on how best to approach the testing

**Comment by gatorsmile** [comment]
See the [code changes](https://github.com/apache/spark/commit/d9c24b3dc5ca64cf8aecebf58f22a3a2e5aab821) I made in my private branch. 

To generate the new result file after you add the new queries into the `.sql` files, you need to run the following command. 
```
SPARK_GENERATE_GOLDEN_FILES=1 build/sbt "sql/test-only *SQLQueryTestSuite"
```

**Comment by selamin** [comment]
@gatorsmile thanks so much for that, i spent over 2 hours getting that sql file correct manually! what was frustrating was after all that i had to revert my changes 😢 

Atleast I know this now so it wasnt a complete waste! Anyways I updated the code to remove the vars

**Comment by gatorsmile** [comment]
LGTM cc @hvanhovell @cloud-fan

**Comment by cloud-fan** [comment]
LGTM

**Comment by samelamin** [comment]
perfect! Can it be merged then?

**Comment by rxin** [comment]
Merging in master.
